composer/callbacks/mlperf.py

```diff
@@ -250,7 +250,9 @@ def _get_dataloader_stats(self, dataloader: Iterable):
             import ffcv  # type: ignore
 
             if isinstance(dataloader, ffcv.loader.Loader):
-                return (dataloader.batch_size, len(dataloader) * dataloader.batch_size)  # type: ignore
+                return (
+                    dataloader.batch_size,  # type: ignore
+                    len(dataloader) * dataloader.batch_size * dist.get_world_size())  # type: ignore
```

Comment:
Can we use this logic for a pytorch dataloader too (instead of just using it for FFCV)?

Reply:
My understanding is `leng(dataloader)` returns number of batches.  So in theory yes, and we can just go back to the original fix of multiplying `num_samples` and `eval_num_samples` by `dist.get_world_size()` in the `_log_dict()`.  However I'm not sure if this will account for when the dataset is pre-sharded (e.g., across nodes).  @hanlint @abhi-mosaic any suggestions on how we should handle this?

Reply:
`len(dataloader)` is always an accurate estimate -- it accounts for the sampler, so it will be "correct" (within rounding error due to distributed or drop last) regardless of whether the dataset is pre sharded or not.

Reply:
Re. @ravi-mosaicml 's comment, whatever logic you decide on, it should work the same for FFCV and a traditional PyTorch dataloader. The sharding of the dataset should not affect things, best I can tell. It's up to the dataset author to make sure that `len(dataset)` is implemented correctly so that `len(dataloader)`, which calls the former, acts correctly.

Slightly tangential, I think it's important to separate:

* how many samples are in the dataset
* how many samples your training loop will consume within one call to `iter(dataloader)`

These values are not the same, because the latter depends on your batch size, the number of devices, and `drop_last` configuration. Which one does MLPerf expect you to report? 

Some illustrative cases:

num_samples = 25, devices=1, batch_size=8, drop_last=True ---> num_samples_consumed=24
num_samples = 25, devices=8, batch_size=8, drop_last=True ---> num_samples_consumed=24
num_samples = 25, devices=1, batch_size=8, drop_last=False ---> num_samples_consumed=25
num_samples = 25, devices=8, batch_size=8, drop_last=False ---> num_samples_consumed=32 ***

Basically things are tricky due to DDP constraints "equal batches per device". I think in the training loop when `drop_last: True` is used, you can get away with the `len(dataloader) * device_batch_size * world_size` hack, but for the eval loop, you've got to be more careful. I think you may again get lucky if `eval_dataset_samples % num_devices == 0`, so like 50000 ImageNet samples divided by 8 GPUs.

Reply:
Ok good point, I looked at NVIDIA's submission and it looks like they're reporting total number of samples in Imagenet.  So we're going to keep using `len(dataloader.dataset)` to compute train/eval_samples for dataloaders that are not FFCV.  For FFCV we will use `len(dataloader.next_traversal_order())` which is what FFCV uses in the length computation:
https://github.com/libffcv/ffcv/blob/main/ffcv/loader/loader.py#L254